R/date_standardisation_helpers.R

```diff
@@ -0,0 +1,405 @@
+
+  first_date_is_charcater <- is.character(first_date)
+  first_date_has_len_1    <- length(first_date) == 1L
+  first_date_has_iso_8601 <- grepl(iso_8601, first_date, perl = TRUE)
+  verdict <- first_date_is_charcater & first_date_has_len_1 &
+    first_date_has_iso_8601
+  if (verdict) {
+    first_date <- as.Date(first_date, "%Y-%m-%d")
+  }
+
+  last_date_is_charcater <- is.character(last_date)
+  last_date_has_len_1    <- length(last_date) == 1L
+  last_date_has_iso_8601 <- grepl(iso_8601, last_date, perl = TRUE)
+  verdict <- last_date_is_charcater & last_date_has_len_1 &
+    last_date_has_iso_8601
+  if (verdict) {
+    last_date <- as.Date(last_date, "%Y-%m-%d")
+  }
+
+  # Set the first date to 50 years before the last date if it's not set
+  if (is.null(first_date) && inherits(last_date, "Date")) {
+    first_date <- min(seq.Date(last_date, length.out = 2L, by = "-50 years"))
+  }
+
+  if (!inherits(first_date, "Date") || !inherits(last_date, "Date")) {
+    stop("first_date and last_date must be Date objects or characters in
+         yyyy-mm-dd format.")
+  }
+  list(first_date, last_date)
+}
+
+#' Convert characters to dates
+#'
+#' @param data the input data frame
+#' @param cols the date column name(s)
+#' @param sep the separator in the date values
+#' @param error_tolerance a number between 0 and 1 indicating the proportion of
+#'     entries which cannot be identified as dates to be tolerated. See the
+#'     `clean_data()` helper for more details
+#'
+#' @return the input data frame where the specified columns have been converted
+#'    into Date.
+#'
+#' @keywords internal
+#' @noRd
+#'
+convert_to_date <- function(data, cols, sep, error_tolerance) {
+  format <- get_format(data, cols, sep)
+  if (!is.null(format)) {
+    data[[cols]] <- as.Date(data[[cols]], format = format)
+  } else {
+    data[[cols]] <- guess_dates(data[[cols]],
+                                error_tolerance = error_tolerance,
+                                check_timeframe = FALSE)
+  }
+  data
+}
+
+#' Guess if a character vector contains Date values, and convert them into Date
+#'
+#' @param data the input data frame
+#' @param error_tolerance a number between 0 and 1 indicating the proportion of
+#'     entries which cannot be identified as dates to be tolerated. See the
+#'     `clean_data()` helper for more details
+#' @param timeframe the expected first and last date. See the `clean_data()`
+#'    helper for more details
+#' @param check_timeframe a logical to determine whether to check if the dates
+#'    fall under the given time frame of not
+#' @param report the object that will contains the report from this operation
+#' @keywords internal
+#' @noRd
+#'
+date_guess_convert <- function(data, error_tolerance,
+                               timeframe, check_timeframe, report) {
+  # guess and convert for column of type character, factor and POSIX
+  col_types <- vapply(data, function(x) class(x)[[1L]], FUN.VALUE = "character")
```

Comment:
This is not a robust way to test inheritance: https://developer.r-project.org/Blog/public/2019/11/09/when-you-think-class.-think-again/index.html

Reply:
I will look into this - I already plan to modify this function such that it will use the output from the `scan_data()` function. That way, only columns that contains a certain amount of `Date` values will be targetted for the conversion.